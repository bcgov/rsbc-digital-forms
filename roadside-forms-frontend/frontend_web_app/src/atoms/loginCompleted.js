import {atom} from 'recoil'

export const loginCompletedAtom = atom({
    key: 'loginCompleted',
    default: false,
  });