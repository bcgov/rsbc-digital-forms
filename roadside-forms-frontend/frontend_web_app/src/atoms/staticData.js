import {atom} from 'recoil'

export const vehicles = atom({
    key:"allVehicles",
    default: [],
  });

export const vehicleStyles = atom({
    key:"allVehicleStyles",
    default: [],
  });

  export const vehicleColours = atom({
    key:"allVehicleColours",
    default: [],
  });

export const provinces = atom({
    key:"allProvinces",
    default: [],
  });

export const permissions = atom({
    key:"allPermissions",
    default: [],
  });

export const jurisdictions = atom({
    key:"allJurisdictions",
    default: [],
  });

export const impoundLotOperators = atom({
    key:"allImpoundLotOperators",
    default: [],
  });

export const countries = atom({
    key:"allCountries",
    default: [],
  });

export const cities = atom({
    key:"allCities",
    default: [],
  });

export const agencies = atom({
    key:"allAgencies",
    default: [],
  });
