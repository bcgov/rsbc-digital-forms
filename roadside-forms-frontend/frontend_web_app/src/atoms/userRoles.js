import {atom} from 'recoil'

const userRolesAtom = atom({
    key: 'userRoles',
    default: [],
  });