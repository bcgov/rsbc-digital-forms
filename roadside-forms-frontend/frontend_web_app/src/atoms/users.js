import { atom } from 'recoil';

export const usersAtom = atom({
    key: 'users',
    default: null
});

export const userAtom = atom({
    key: 'user',
    default: []
});