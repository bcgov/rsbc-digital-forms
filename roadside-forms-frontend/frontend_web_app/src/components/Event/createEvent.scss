.form-id-border{
    border: 1px dashed black;
    color: black;
    background-color: #F2F2F2;
    border-radius: 0.2rem;
    text-align: center;
    padding-top: 40px;
}

