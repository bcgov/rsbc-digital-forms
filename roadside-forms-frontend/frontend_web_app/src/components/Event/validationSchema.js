import * as Yup from 'yup';

const vehicleNotImpoundedValidation = (selectedValue) => {
  return Yup.string().test('vehicleNotImpounded', 'This field is required', function(value) {
    const { createError, path, options } = this;

    if (selectedValue === 'Yes' && !value) {
      return createError({ path, message: options.message });
    }

    return true;
  });
};

const vehicleImpoundedValidation = (selectedValue) => {
  return Yup.string().test('vehicleImpounded', 'This field is required', function(value) {
    const { createError, path, options } = this;

    if (selectedValue === 'Yes' && !value) {
      return createError({ path, message: options.message });
    }

    return true;
  });
};

const typeOfProhibitionAlcoholValidation = (selectedValue) => {
  return Yup.string().test('typeOfProhibitionAlcohol', 'This field is required', function(value) {
    const { createError, path, options } = this;

    if (selectedValue === 'alcohol' && !value) {
      return createError({ path, message: options.message });
    }

    return true;
  });
};

const typeOfProhibitionDrugsValidation = (selectedValue) => {
  return Yup.string().test('typeOfProhibitionDrugs', 'This field is required', function(value) {
    const { createError, path, options } = this;

    if (selectedValue === 'drugs' && !value) {
      return createError({ path, message: options.message });
    }

    return true;
  });
};

const releasedToDriverValidation = (selectedValue) => {
  return Yup.string().test('releasedToDriver', 'This field is required', function(value) {
    const { createError, path, options } = this;

    if (selectedValue === 'released' && !value) {
      return createError({ path, message: options.message });
    }

    return true;
  });
};

const prohibitionValidation = (yesSeleted) => {
  return Yup.string().test('prohibition', 'This field is required', function(value) {
    const { createError, path, options } = this;

    if (yesSeleted && !value) {
      return createError({ path, message: options.message });
    }

    return true;
  });
};

export const validationSchema = Yup.object().shape({
  //common form fields validation
  "last-name": Yup.string().required('Last Name is required'),
  "address": Yup.string().required('Address is required'),
  "city": Yup.string().required('City is required'),
  "prov-state": Yup.string().required('Prov / State is required'),
  "officer-lastname": Yup.string().required('Last Name is required'),
  "officer-prime-id": Yup.string().required('PRIME ID is required'),
  "officer-agency": Yup.string().required('Agency is required'),
  "phone": Yup.string().matches(/^\d{3}-\d{3}-\d{4}$/, 'Phone number format ###-###-####'),
  "dob": Yup.string()
  .nullable()
  .test('dob-validation', 'Invalid Date of Birth', (dob) => {
    if (!dob) return true;

    const dateRegex = /^(19|20)\d{2}(0[1-9]|1[0-2])(0[1-9]|[12]\d|3[01])$/;
    if (!dateRegex.test(dob)) return false;

    const year = Number(dob.substring(0, 4));
    const month = Number(dob.substring(4, 6));
    const day = Number(dob.substring(6, 8));

    const currentDate = new Date();
    const inputDate = new Date(year, month - 1, day);

    // Check if the input date is valid and within the desired range
    return (
      inputDate.getFullYear() === year &&
      inputDate.getMonth() === month - 1 &&
      inputDate.getDate() === day &&
      inputDate <= currentDate &&
      year >= 1900
    );
  }),
  "vin-number": Yup.string().max(20, 'VIN must be 20 characters or less'),
  "nsc-number": Yup.string().max(14, 'NSC no. must be 14 characters or less'),
  //24 Hour Fields validation
  'vehicle-impounded': Yup.string().test('vehicleImpoundment', 'Vehicle impoundment is required when 24-hour is selected', function (value) {
    const twentyFourHourValue = this.parent['24Hour'];
    
    if (twentyFourHourValue && !value) {
      return this.createError({
        path: 'vehicle-impounded',
        message: 'Vehicle impoundment is required',
      });
    }

    return true;
  }),
  'key-location': vehicleImpoundedValidation(Yup.ref('vehicle-impounded')),
  'ILO-name': vehicleImpoundedValidation(Yup.ref('vehicle-impounded')),
  'ILO-address': vehicleImpoundedValidation(Yup.ref('vehicle-impounded')),
  'ILO-city': vehicleImpoundedValidation(Yup.ref('vehicle-impounded')),
  'ILO-phone': vehicleImpoundedValidation(Yup.ref('vehicle-impounded')),
  'reason-for-not-impounding': vehicleNotImpoundedValidation(Yup.ref('vehicle-impounded')),
  "type-of-prohibition": prohibitionValidation(Yup.ref('24Hour')),
  "offence-address": prohibitionValidation(Yup.ref('24Hour')),
  "offence-city": prohibitionValidation(Yup.ref('24Hour')),
  "offence-agency-file": prohibitionValidation(Yup.ref('24Hour')),
  'date-of-driving': Yup.date()
  .max(new Date(), 'Date of driving cannot be a future date')
  .nullable()
  .test('prohibition', 'Date of driving is required when 24-hour is selected', function(value) {
    const twentyFourHourValue = this.parent['24Hour'];

    if (twentyFourHourValue && !value) {
      return this.createError({
        path: 'date-of-driving',
        message: 'Date of driving is required',
      });
    }

    return true;
  }),
  "time-of-driving": prohibitionValidation(Yup.ref('24Hour')),
  "vehicle-released-to": releasedToDriverValidation(Yup.ref('reason-for-not-impounding')),
  "date-released": Yup.date()
  .max(new Date(), 'Date of release cannot be a future date')
  .nullable()
  .test('prohibition', 'Date of release is required when release is selected', function(value) {
    const selectedValue = this.parent['reason-for-not-impounding'];

    if (selectedValue === 'released' && !value) {
      return this.createError({
        path: 'date-of-release',
        message: 'Date of release is required',
      });
    }

    return true;
  }),
  "time-released": releasedToDriverValidation(Yup.ref('reason-for-not-impounding')),
  "test-used-alcohol": typeOfProhibitionAlcoholValidation(Yup.ref('type-of-prohibition')),
  "BAC-result": typeOfProhibitionAlcoholValidation(Yup.ref('type-of-prohibition')),
  'ASD-expiry-date': Yup.date()
  .nullable()
  .test('ASD', 'ASD expiry date is required when ASD is is selected', function(value) {
    const selectedValue = this.parent['test-used-alcohol'];

    if (selectedValue === 'alco-sensor' && !value) {
      return this.createError({
        path: 'ASD-expiry-date',
        message: 'This field is required',
      });
    }

    return true;
  }),
  'alcohol-test-result': Yup.date()
  .nullable()
  .test('Instrument', 'Test result is required when Instrument is is selected', function(value) {
    const selectedValue = this.parent['test-used-alcohol'];

    if (selectedValue === 'instrument' && !value) {
      return this.createError({
        path: 'instrument',
        message: 'This field is required',
      });
    }

    return true;
  }),
  "test-used-drug": typeOfProhibitionDrugsValidation(Yup.ref('type-of-prohibition')),
});
