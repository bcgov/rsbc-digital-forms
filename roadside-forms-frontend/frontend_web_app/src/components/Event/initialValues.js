import { useRecoilValue } from 'recoil';
import { userAtom } from '../../atoms/users';

export const InitialValues = () => {
    const user = useRecoilValue(userAtom);

    return {
        //common fields
        "IRP": false,
        "VI": false,
        "24Hour": false,
        "12Hour": false,
        "drivers-licence-jurisdiction":{ value: 'BC', label: 'BRITISH COLUMBIA' },
        "drivers-number":'',
        "last-name":'',
        "given-name":'',
        "dob": '',
        "address":'',
        "city":'',
        "phone": '',
        "prov-state":{ value: 'BC', label: 'BRITISH COLUMBIA' },
        "postal-code":'',
        "vehicle-jurisdiction":{ value: 'BC', label: 'BRITISH COLUMBIA' },
        "plate-number": '',
        "registration-number": '',
        "vehicle-year": '',
        "vehicle-make-model": '',
        "vehicle-style": '',
        "vehicle-colour": [],
        "nsc-prov-state":{ value: 'BC', label: 'BRITISH COLUMBIA' },
        "vin-number": '',
        "nsc-number": '',
        "owned-by-corp" : false,
        "corp-name": '',
        "owner-last-name":'',
        "owner-first-name": '',
        "registered-owner-address": '',
        "registered-owner-phone": '',
        "registered-owner-city": '',
        "registered-owner-prov-state": {},
        "registered-owner-postal": '',
        "officer-lastname":user.last_name || '',
        "officer-prime-id":user.badge_number || '',
        "officer-agency":user.agency || '',
        //24-hour form fields
        "vehicle-impounded":'',
        "key-location": '',
        "reason-for-not-impounding": '',
        'ILO-options': '',
        "ILO-name": '',
        "ILO-address":'',
        "ILO-city":'',
        "ILO-phone":'',
        "vehicle-released-to": '',
        "date-released": '',
        "time-released": '',
        "type-of-prohibition": '',
        "offence-address":'',
        "offence-city":'',
        "offence-agency-file":'',
        "date-of-driving":'',
        "time-of-driving": '',
        "witnessed-by-officer":false,
        "admission-by-driver":false,
        "independent-witness":false,
        "video-surveillance":false,
        "other-reasonable-ground":false,
        "prescribed-device":'',
        "test-used-alcohol": '',
        "ASD-expiry-date": '',
        "alcohol-test-result": '',
        "BAC-result": '',
        "test-used-drug": '',
        "THC":false,
        "Cocaine":false,
    }
}