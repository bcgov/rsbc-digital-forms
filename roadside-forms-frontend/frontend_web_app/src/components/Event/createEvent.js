import React, { useState, useEffect } from 'react';
import { Formik, Form} from 'formik';
import { Checkbox } from '../common/Checkbox/checkbox';
import { validationSchema } from './validationSchema';
import { Button } from '../common/Button/Button';
import { DriverInfo } from '../CommonForm/driverInfo';
import { InitialValues } from './initialValues';
import { VehicleInfo } from '../CommonForm/vehicleInfo';
import { OfficerInfo } from '../CommonForm/officerInfo';
import { RegisteredOwnerInfo } from '../CommonForm/registeredOwnerInfo';
import { useRecoilValue } from 'recoil';
import { staticResources } from '../../utils/helpers';
import { useNavigate } from 'react-router-dom';
import './createEvent.scss';

export const CreateEvent = () => {
    const vehicleStylesAtom = useRecoilValue(staticResources["vehicle_styles"]);
    const vehicleColoursAtom = useRecoilValue(staticResources["vehicle_colours"]);
    const jurisdictionsAtom = useRecoilValue(staticResources["jurisdictions"]);
    const provincesAtom = useRecoilValue(staticResources["provinces"]);
    const vehiclesAtom= useRecoilValue(staticResources["vehicles"]);
    const [jurisdictions, setJurisdictions] = useState([]);
    const [provinces, setProvinces] = useState([]);
    const [vehicleStyles, setVehicleStyles] = useState([]);
    const [vehicles, setVehicles] = useState([]);
    const [vehicleColours, setVehicleColours] = useState([]);
    const navigate = useNavigate();

    useEffect(() => {
        setJurisdictions(
            jurisdictionsAtom.map((each) => ({ label: each.objectDsc, value: each.objectCd }))
        );
        setProvinces(
            provincesAtom.map((each) => ({ label: each.objectDsc, value: each.objectCd }))
        );
        setVehicleStyles(
            vehicleStylesAtom.map((each) => ({ label: each.name, value: each.code }))
        );
        setVehicleColours(
            vehicleColoursAtom.map((each) => ({ label: each.display_name, value: each.code }))
        );
        setVehicles(
            vehiclesAtom.map((each) => ({ label: each.search, value: each.search }))
        );
      }, [
        vehicleStylesAtom,
        jurisdictionsAtom,
        provincesAtom,
        vehiclesAtom,
        vehicleColoursAtom
    ]);

    console.log("vehicleColours", vehicleColours);

    const generateYearOptions = () => {
        const currentYear = new Date().getFullYear();
        const startYear = 1991;
        const years = [];
      
        for (let year = startYear; year <= currentYear; year++) {
          years.push({ value: year, label: year.toString() });
        }
      
        return years;
      };

    const onSubmit = (values, { setSubmitting }) => {
        console.log('submit data', values);
        setSubmitting(false);
    };

    const handleGoBack = () => {
        navigate(-1);
      };

    return (
        <div className='text-font'>
            <div className='m-4'>
                <Button onClick={handleGoBack}  primary size='large' label="Save & Return to Main Menu"/>
            </div>
            <div className="outline">
            <Formik initialValues={InitialValues()} validationSchema={validationSchema} onSubmit={onSubmit}>
                {({ isSubmitting }) => (
                    <Form>
                        <div className='row mt-2'>
                            <div className='col-sm-4 left checkboxs'>
                                <h4>Documents to Generate</h4>
                                <Checkbox name="IRP">Immediate Roadside Prohibition</Checkbox>
                                <Checkbox name="VI">Vehicle Impound</Checkbox>
                                <Checkbox name="24Hour">24-hour Driving Prohibition</Checkbox>
                                <Checkbox name="12Hour">12-hour Driving Prohibition</Checkbox>
                            </div>
                            <div className='col-sm-4 form-id-border'>
                                <h5>IRP number: 21-9876540</h5>
                                <h5>VI number: 22-1234560</h5>
                            </div>
                            <div className='col-sm-4 time-of-completion center mt-5'>
                                <span>Estimated time to complete:</span>
                                <h5>12 minutes</h5>
                            </div>
                        </div>
                        <div className='common-fields'>
                            <DriverInfo jurisdictions={jurisdictions} provinces={provinces}/>
                            <VehicleInfo vehicleColours={vehicleColours} years={generateYearOptions()} provinces={provinces} jurisdictions={jurisdictions} vehicles={vehicles} vehicleStyles={vehicleStyles}/>
                            <RegisteredOwnerInfo provinces={provinces}/>
                            <OfficerInfo></OfficerInfo>
                        </div>
                        <div className='right'>
                            <Button primary size='large' label="Next: Preview" type="submit" disabled={isSubmitting}/>
                        </div>
                    </Form>
                )}
            </Formik>
            </div>
        </div>
    );
}