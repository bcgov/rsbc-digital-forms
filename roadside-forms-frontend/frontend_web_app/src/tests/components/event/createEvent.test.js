import React from 'react';
import { render, act } from '@testing-library/react';
import { RecoilRoot } from 'recoil';
import { CreateEvent } from '../../../components/Event/createEvent';
import { vehicleStyles, jurisdictions, provinces, vehicles, vehicleColours, cities, impoundLotOperators } from '../../../atoms/staticData';

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: jest.fn(),
  }));
describe('CreateEvent component', () => {
    test('renders create event component correctly', async () => {
    let container; 

    await act(async () => {
      ({ container } = render( 
        <RecoilRoot
          initializeState={(snap) => {
            snap.set(vehicleStyles, [  {
                "code": "2DR",
                "name": "2-DOOR "
            },
            {
                "code": "3DR",
                "name": "3-DOOR "
            }]);
            snap.set(jurisdictions, [{
                "objectCd": "AL",
                "objectDsc": "ALABAMA"
            },
            {
                "objectCd": "AK",
                "objectDsc": "ALASKA"
            }]);
            snap.set(provinces, [{
                "objectCd": "AK",
                "objectDsc": "ALASKA"
            },
            {
                "objectCd": "ALTA",
                "objectDsc": "ALBERTA"
            }]);
            snap.set(vehicles, [{
                "mk": "AC-TEST",
                "search": "A C (GREAT BRITAIN)",
                "md": ""
            },
            {
                "mk": "TEST",
                "search": "A C (GREAT BRITAIN) - 3000 ME",
                "md": "300"
            }]);
            snap.set(vehicleColours, [
                {"code": 'test', 
                "display_name": 'test'}
            ]);
            snap.set(cities, [{
                "objectCd": "TEST",
                "objectDsc": "10 TEST HOUSE"
            },
            {
                "objectCd": "OHTM",
                "objectDsc": "10390 TEST ST"
            }]);
            snap.set(impoundLotOperators, [
            {
                "name": "TSET",
                "lot_address": "7 TEST ST",
                "city": "TEST",
                "phone": "250-003-73654"
            },
            {
                "name": "TSET",
                "lot_address": "7 TEST ST",
                "city": "TEST",
                "phone": "250-003-73654"
            }]);
          }}
        >
          <CreateEvent />
        </RecoilRoot>
      ));
    });

    expect(container).toMatchSnapshot();
});
});
