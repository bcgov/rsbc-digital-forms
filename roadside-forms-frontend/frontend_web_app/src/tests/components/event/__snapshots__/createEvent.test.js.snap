// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateEvent component renders create event component correctly 1`] = `
<div>
  <div
    class="text-font"
    id="event-container"
  >
    <div
      class="Toastify"
    />
    <div
      class="flex"
    >
      <div
        class="m-4"
        id="button-container"
      >
        <button
          class="btn btn-primary"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ArrowBackIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20z"
            />
          </svg>
          Â  Return to Main Menu
        </button>
      </div>
    </div>
    <div
      class="outline"
    >
      <form
        action="#"
      >
        <div
          class="row mt-2"
        >
          <div
            class="col-sm-4 left checkboxs"
          >
            <h3>
              Please select one or more forms:
            </h3>
            <h5>
              <div>
                <label>
                  <input
                    class="mr-1"
                    id="VI"
                    name="VI"
                    type="checkbox"
                    value="false"
                  />
                  <span
                    style="margin-left: 5px;"
                  >
                    Vehicle Impound
                  </span>
                </label>
              </div>
            </h5>
            <h5>
              <div>
                <label>
                  <input
                    class="mr-1"
                    id="TwentyFourHour"
                    name="TwentyFourHour"
                    type="checkbox"
                    value="false"
                  />
                  <span
                    style="margin-left: 5px;"
                  >
                    24-hour Driving Prohibition
                  </span>
                </label>
              </div>
            </h5>
            <h5>
              <div>
                <label>
                  <input
                    class="mr-1"
                    id="TwelveHour"
                    name="TwelveHour"
                    type="checkbox"
                    value="false"
                  />
                  <span
                    style="margin-left: 5px;"
                  >
                    12-hour Driving Suspension
                  </span>
                </label>
              </div>
            </h5>
          </div>
          <div
            class="col-sm-4 form-id-border"
          />
        </div>
        <div
          class="flex"
          id="button-container"
        />
      </form>
    </div>
  </div>
</div>
`;
